Plans 2023: Migrate p2 plans to new pricing grid

jeyip opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/77652

## Proposed Changes

* For the p2 flow plans page, replace the outdated pricing page with the 2023 pricing page `PlansFeaturesMain` component

**Note:**

We will be removing outdated references ( examples below ) to P2 plans in a separate PR

https://github.com/Automattic/wp-calypso/blob/ba8c3ff726444513308b8c91b740fbdee9df818c/client/my-sites/plan-features/actions.jsx#L67

https://github.com/Automattic/wp-calypso/blob/ba8c3ff726444513308b8c91b740fbdee9df818c/client/my-sites/plan-features/header.jsx#L42-L44, etc. )

We're also focusing less on perfecting the design and more on removing custom layout and styling of the previous p2 workspaces plans page. This is because the feature is in maintenance mode. Context here p9lV3a-6EY-p2

## Screenshots

### Before

<img width="857" alt="Screenshot 2023-10-13 at 8 40 10 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/9c1be856-3d11-4085-9b55-4ff98cde5d8a">

### After

<img width="1174" alt="Screenshot 2023-10-14 at 12 22 34 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/b22195ce-0c3b-4d00-b49d-eef1c8c0e6f6">

<img width="992" alt="Screenshot 2023-10-14 at 12 28 36 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/de9fab7f-1e63-4f7c-b838-ba31199ff474">

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Create a new p2 site by visiting calypso.localhost:3000/start/p2 and following onboarding instructions
* Note the site slug that's created for your new p2 workspace
* Navigate directly to calypso.localhost:3000/plans/{P2_WORKSPACE_SITE_SLUG}
* Verify that the P2+ plan can still be purchased
* Verify that the "Your plan" badge highlights the correct plan
* Verify that, when the current plan is the p2 free plan, that the "Manage add-ons" button redirects to the /add-ons page in Calypso admin
* Verify that everything behaves as expected in the mobile view

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

jeyip committed:
Initial migration to 2023 plans grid for p2 plans

jeyip committed:
Fix upgrade button styling

jeyip committed:
Fix p2 plans storage labels

jeyip committed:
Make p2 plan description into tagline

Currently, descriptions aren't rendered in the new 2023 plans grid. We place the description into the plan tagline instead

jeyip committed:
Remove unnecessary commented code

jeyip committed:
Fix yearlyVariantPricing runtime error

yearlyVariantPlanSlug can not, in fact, be an empty string. Otherwise usePricingMetaForGridPlans > usePricedAPIPlans > useProductIds > applyTestFiltersToPlansList throws a runtime error

Fix TranslateResult type errors

Fix is-plan.js unit test

Fix plan-lookup test

jeyip commented on client/my-sites/plans-grid/components/billing-timeframe.tsx:
```diff
@@ -35,13 +35,15 @@ function usePerMonthDescription( { planSlug }: { planSlug: PlanSlug } ) {
-		withoutProRatedCredits: true,
-		storageAddOns: storageAddOnsForPlan,
-	} )?.[ yearlyVariantPlanSlug ];
+	const yearlyVariantPlanSlug = getPlanSlugForTermVariant( planSlug, TERM_ANNUALLY );
```
`yearlyVariantPlanSlug` cannot, in fact, be an empty string. Otherwise `usePricingMetaForGridPlans` > `usePricedAPIPlans` > `useProductIds` > `applyTestFiltersToPlansList` throws a runtime error

chriskmnds replied:
interesting

jeyip commented on client/my-sites/plans-grid/hooks/npm-ready/data-store/use-plans-from-types.ts:
```diff
@@ -6,26 +6,36 @@ import {
 			: { type, term };
+
+		if ( intent === 'plans-p2' ) {
+			planQuery[ 'group' ] = GROUP_P2;
```
**Note:**

All free plans have identical types (  `TYPE_PLAN_FREE` ) and terms. To query the correct P2 free plan, we add a `group` query property

jeyip committed:
Fix current plan logic for plan p2 free

jeyip committed:
Remove unnecessary CTA plan p2 logic

jeyip commented on client/my-sites/plans/p2-plans-main.jsx:
```diff
@@ -1,30 +1,27 @@
 
 		return (
 			<>
 				<QueryPlans />
```
Unsure if these queries are necessary 🤔 If anyone happens to know, please feel free to chime in

chriskmnds replied:
They are probably not needed as the `plans-features-main` queries the same there. I think I'd keep them just to be safe. `plans-features-main` will be going through changes and might affect which queries are called?

jeyip replied:
Makes sense 🙂

In that case I'll leave them for now

jeyip commented:
**Note:**

We're focusing less on perfecting the design and more on removing custom layout and styling of the previous p2 workspaces plans page. This is because the feature is in maintenance mode. Context here p9lV3a-6EY-p2

chriskmnds commented on client/my-sites/plans-grid/hooks/npm-ready/data-store/is-current-plan.ts:
```diff
@@ -0,0 +1,8 @@
+import { PLAN_FREE, isFreePlan, type PlanSlug } from '@automattic/calypso-products';
+
+export function isCurrentPlan( currentSitePlanSlug: string, planSlug: PlanSlug ): boolean {
```
Ideally, we ought to only add hooks under `plans-grid/hooks` folder. Maybe move this into `plans-grid/lib`?

jeyip replied:
My mistake -- I saw the `is-popular-plan` helper in `hooks/npm-ready/data-store`. I'm happy to scoot this into `plans-grid/lib` 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/82786/commits/0b8125aa2c5a617c46fa90ea57ea03eda73ca9f7

chriskmnds replied:
Ah maybe you were correct to put it under there then! But no need to change anything back if you are ready to ship. There's some way before we extract and organise all these in a data store.

chriskmnds commented on client/my-sites/plans-grid/hooks/npm-ready/data-store/is-current-plan.ts:
```diff
@@ -0,0 +1,8 @@
+import { PLAN_FREE, isFreePlan, type PlanSlug } from '@automattic/calypso-products';
+
+export function isCurrentPlan( currentSitePlanSlug: string, planSlug: PlanSlug ): boolean {
```
Something about this signature though doesn't seem correct to me. Maybe renaming to `isSamePlan` might be more precise, as that's what it does technically?

jeyip replied:
Oh that's interesting 🤔 This wasn't even on my radar, but yeah I agree that `isSamePlan` would be better. 

Out of curiosity, how did you sense that it seemed off? Was it the redundant usage of the term "current"?

jeyip replied:
[dbb0dea](https://github.com/Automattic/wp-calypso/pull/82786/commits/dbb0deaf7576e59cda62a67d14aa2d8fbf00b1b0)

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/82786/commits/dbb0deaf7576e59cda62a67d14aa2d8fbf00b1b0

chriskmnds replied:
Not sure. I think so. It seemed a little off, hard to describe in words maybe? Not saying the alternative is better btw. It's all a bit convoluted and relative I think. Maybe the original might make more sense with a good description 🤷

chriskmnds reviewed (approved):
Nicely done! Tests out per the instructions. I don't notice any obvious regressions (outside of the points below). I left some comments in the code.

- Not sure why this happens, but the styling on the sticky bar in `/plans/[ any site ]` seems broken on the live image:

<img width="400" alt="Screenshot 2023-10-17 at 2 46 50 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/6d4ff62d-ac58-4cb4-928f-425522d27c0a">

- Also the free plan in comparison grid like you mentioned:

<img width="400" alt="Screenshot 2023-10-17 at 2 51 02 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/ce89a3e6-4d05-4e8a-aad8-4c03f9e6f98a">

Maybe a rebase might do? I recall we "fixed" (more like patched) the comparison grid last week 🤷‍♂️

chriskmnds commented:
^ Happy to test more if you apply any changes post-rebase

jeyip commented:
Hmmm...currently doing some local dev debugging. I'm seeing lots of runtime errors about our bundling process and have no clue why. Will continue working on this PR when I've figured out what's happening

jeyip committed:
Migrate isCurrentPlan into plans-grid/lib

jeyip committed:
Rename isCurrentPlan to isSamePlan

jeyip commented:
It looks like rebasing solved both issues :D

chriskmnds commented:
Thanks for the follow-ups. On hindsight, my comments may have been a bit misleading 😬  but as long as it resonates. All good either way.

jeyip commented:
Holy cow what a complex adventure. I figured out the issue https://github.com/Automattic/wp-calypso/pull/82786#issuecomment-1767561238. I'm determining a fix now

jeyip committed:
Fix is current plan property

jeyip committed:
Fix layout and styling error

We're replacing the legacy PlanFeatures component with PlanFeaturesMain. There were, however, styles being loaded from the PlanFeatures component whenever users visited the /plans page, even if the PlanFeatures component itself wasn't being rendered.

Now that we're removing the old PlanFeatures component, said styles are no longer loaded, so we ensure that a critical css transform is still applied with the new PlanFeaturesMain component

southp commented:
![](https://live.staticflickr.com/1003/1449650710_61aa5906d5.jpg)
